File: cpp/src/img_acq/frame_grabber.cpp
#include <chrono>
#include <mutex>
#include <opencv2/core/mat.hpp>
#include <opencv2/videoio.hpp>
#include <setdetect/img_acq/frame_grabber.hpp>
#include <stdexcept>
#include <string>
#include <thread>
#include <utility>

using namespace std::chrono;

namespace setdetect::img_acq {

FrameGrabber::FrameGrabber(const std::string& source, FrameCallback on_frame, double target_fps)
    : on_frame_(std::move(on_frame)), target_fps_(target_fps) {
    if (!cap_.open(source)) {
        throw std::runtime_error("Could not open source: " + source);
    }
    if (target_fps_ <= 0.0) {
        target_fps_ = cap_.get(cv::CAP_PROP_FPS);
    }
    if (target_fps_ <= 0.0) {
        target_fps_ = 30.0;
    }
}

FrameGrabber::~FrameGrabber() { stop(); }

void FrameGrabber::start() { thread_ = std::thread(&FrameGrabber::run, this); }

void FrameGrabber::stop() {
    stop_.store(true);
    if (thread_.joinable()) {
        thread_.join();
    }
    cap_.release();
}

void FrameGrabber::run() {
    auto period = duration<double>(1.0 / target_fps_);
    auto next_tick = steady_clock::now();

    while (!stop_.load()) {
        cv::Mat frame;
        if (!cap_.read(frame)) {
            break;
        }

        Frame const f{std::move(frame), duration<double>(steady_clock::now().time_since_epoch()).count(), frame_id_++};
        on_frame_(f);

        next_tick += duration_cast<steady_clock::duration>(period);
        auto now = steady_clock::now();
        if (next_tick > now) {
            std::this_thread::sleep_until(next_tick);
        } else {
            next_tick = now;
        }
    }
    stop_.store(true);
}

AsyncDispatcher::AsyncDispatcher(FrameCallback handler) : handler_(std::move(handler)) {
    worker_ = std::thread(&AsyncDispatcher::loop, this);
}

AsyncDispatcher::~AsyncDispatcher() { stop(); }

void AsyncDispatcher::operator()(const Frame& frame) {
    std::lock_guard<std::mutex> const guard(lock_);
    slot_ = frame;
    has_slot_.store(true);
}

void AsyncDispatcher::stop() {
    stop_.store(true);
    if (worker_.joinable()) {
        worker_.join();
    }
}

void AsyncDispatcher::clear() {
    std::lock_guard<std::mutex> const guard(lock_);
    has_slot_.store(false);
}

void AsyncDispatcher::loop() {
    while (!stop_.load()) {
        if (has_slot_.load()) {
            Frame frame;
            {
                std::lock_guard<std::mutex> const guard(lock_);
                frame = slot_;
                has_slot_.store(false);
            }
            handler_(frame);
        }
        std::this_thread::sleep_for(std::chrono::milliseconds(1));
    }
}

} // namespace setdetect::img_acq

File: py/src/setdetect/util/frame_grabber.py
import threading
import time
import typing as tp
from collections.abc import Callable
from dataclasses import dataclass

import cv2


@dataclass
class Frame:
    data: tp.Any
    timestamp: float
    frame_id: int


FrameCallback = Callable[[Frame], None]


class FrameGrabber:
    """
    Push-based grabber: mimics a camera SDK's event model.
    The registered callback is invoked directly from the
    acquisition thread.
    """

    def __init__(self, source, on_frame: FrameCallback, target_fps: float | None = None):
        self.cap = cv2.VideoCapture(source)
        if not self.cap.isOpened():
            raise RuntimeError(f"Could not open source: {source}")

        self.on_frame = on_frame
        self.target_fps = target_fps or self.cap.get(cv2.CAP_PROP_FPS) or 30.0
        self._stop = threading.Event()
        self._frame_id = 0
        self._thread = threading.Thread(target=self._run, daemon=True)

    def start(self):
        self._thread.start()
        return self

    def _run(self):
        period = 1.0 / self.target_fps
        next_tick = time.monotonic()

        while not self._stop.is_set():
            ok, frame = self.cap.read()
            if not ok:
                break

            f = Frame(data=frame, timestamp=time.monotonic(), frame_id=self._frame_id)
            self._frame_id += 1

            self.on_frame(f)

            next_tick += period
            sleep_time = next_tick - time.monotonic()
            if sleep_time > 0:
                time.sleep(sleep_time)
            else:
                next_tick = time.monotonic()

        self._stop.set()

    def stop(self):
        self._stop.set()
        self._thread.join(timeout=1.0)
        self.cap.release()


class AsyncDispatcher:
    def __init__(self, handler: FrameCallback, drop_policy="drop_oldest"):
        self.handler = handler
        self.drop_policy = drop_policy
        self._slot: Frame | None = None
        self._lock = threading.Lock()
        self._new_frame = threading.Event()
        self._stop = threading.Event()
        self._worker = threading.Thread(target=self._loop, daemon=True)
        self._worker.start()

    def __call__(self, frame: Frame):
        with self._lock:
            if self._slot is not None and self.drop_policy == "drop_newest":
                return
            self._slot = frame
        self._new_frame.set()

    def _loop(self):
        while not self._stop.is_set():
            if self._new_frame.wait(timeout=0.1):
                self._new_frame.clear()
                with self._lock:
                    frame, self._slot = self._slot, None
                if frame is not None:
                    self.handler(frame)

    def stop(self):
        self._stop.set()
        self._worker.join(timeout=2.0)

    def clear(self):
        with self._lock:
            self._slot = None
        self._new_frame.clear()
